File: app/src/main/java/com/example/oceo/speedread/EPubLibUtil.java
package com.example.oceo.speedread;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;
import android.widget.Toast;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import nl.siegmann.epublib.domain.Book;
import nl.siegmann.epublib.domain.Guide;
import nl.siegmann.epublib.domain.MediaType;
import nl.siegmann.epublib.domain.Resource;
import nl.siegmann.epublib.domain.Resources;
import nl.siegmann.epublib.domain.Spine;
import nl.siegmann.epublib.domain.SpineReference;
import nl.siegmann.epublib.domain.TOCReference;
import nl.siegmann.epublib.domain.TableOfContents;
import nl.siegmann.epublib.epub.EpubReader;
import nl.siegmann.epublib.service.MediatypeService;

public class EPubLibUtil {
    private String testBook1Path = "storage/emulated/0/Books/MoonReader/Brandon Sanderson - Oathbringer_ Book Three of the Stormlight Archive-Tor Books (2017).epub";
    static MediaType[] bitmapTypes = {MediatypeService.PNG, MediatypeService.GIF, MediatypeService.JPG};


    public static Book getBook(String fName) {
        File file = new File(fName);
        Book book = null;

        try {
            InputStream epubInputStream = new FileInputStream(file.toString());
            Log.d("what is it", file.toString());
            if (file.toString().contains(".epub")) {
                book = (new EpubReader()).readEpub(epubInputStream);
            } else {
                Log.d("GetBook", "Not an Epub");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return book;
    }

    /*
       trying to display images
    */
    public static Bitmap getBitmapFromResources(List<Resource> resourcez, String imgHref, Book book) {
        List<Resource> resources = book.getResources().getResourcesByMediaTypes(bitmapTypes);
        byte[] data = "holder".getBytes();
        for (int ii = 0; ii < resources.size(); ii++) {
            String z = resources.get(ii).getHref();
            Log.d("given: ", imgHref);
            Log.d("the href: ", z);
            Log.d("contains", String.valueOf(imgHref.contains(z)));
            Log.d("contains", String.valueOf(z.contains(imgHref)));
            // it isnt clear which one will be larger
            // oathbringer chpt 4 is an example of the first condition
            // hyperian prologue is an example of seocnd
            if (imgHref.contains(z) || z.contains(imgHref)) {
//                Log.i("livi", z);
                try {
                    data = resources.get(ii).getData();
                } catch (IOException e) {
                    // TODO Auto-generated catch block
                    e.printStackTrace();
                }
                break;
            }
        }
        //
        Bitmap bm = BitmapFactory.decodeByteArray(data, 0, data.length);
        return bm;
    }


    public static void getContents(Book book) {
        List<Resource> contents = book.getContents();
    }

    public static void exploreResources(Book book) {
        //TODO
        Log.d("explorerResources", "start");
        Resources resource = book.getResources();
        Log.d("explorerResources", "done");

    }


    public static List<TOCReference> exploreTOC(Book book) {
        Log.d("explorerTOC", "start");
        TableOfContents toc = book.getTableOfContents();
//        List<Resource> uniqueRes = toc.getAllUniqueResources();
        List<TOCReference> tocReferences = toc.getTocReferences();
        Log.d("explorerTOC", "done");
        return tocReferences;
    }


    public static void exploreGuide(Book book) {
        Log.d("exploreGuide", "start");
        Guide guide = book.getGuide();
        Log.d("exploreGuide", "done");
    }


    public static void exploreSpine(Book book) {
        Log.d("exploreSpine", "start");
        Spine spine = book.getSpine();
        Log.d("exploreSpine", "done");
    }

    public static void unZipEPUB() {
        //TODO
//        FileUtils.unzipFile(mFilePath, Constant.PATH_EPUB + "/" + mFileName);
    }

    public static int mapSpineToTOC(String spineID, ArrayList<String> tocResList) {
        int i = 0;
        while (i < tocResList.size()) {
            if (spineID.equals(tocResList.get(i))) {
                return i;
            }
            i++;
        }
        return -1;
    }


    public static int mapTOCToSpine(Book book, String tocID) {
        List<SpineReference> spineRefs = book.getSpine().getSpineReferences();
        int i = 0;
        while (i < spineRefs.size()) {
            String spineID = spineRefs.get(i).getResource().getId();
            if (spineID.equals(tocID)) {
                return i;
            }
            i++;
        }
        return -1;
    }


    /*

     */
    public static ArrayList<String> getTOCResourceIds(List<TOCReference> tocReferences, int depth, ArrayList<String> toc) {
        // TODO examine
        if (tocReferences == null) {
            return toc;
        }
        for (TOCReference tocReference : tocReferences) {
            StringBuilder tocString = new StringBuilder();
            for (int i = 0; i < depth; i++) {
                tocString.append("\t");
            }
            tocString.append(tocReference.getTitle());
//            Log.i("epublib", tocString.toString());
//            toc.add(tocString.toString());
            toc.add(tocReference.getResource().getId()); // resource Id
            getTOCResourceIds(tocReference.getChildren(), depth + 1, toc);
        }
        return toc;
    }

    public static ArrayList<String> getTOCTitles(List<TOCReference> tocReferences, int depth, ArrayList<String> toc) {
        // TODO examine
        if (tocReferences == null) {
            return toc;
        }
        for (TOCReference tocReference : tocReferences) {
            StringBuilder tocString = new StringBuilder();
            for (int i = 0; i < depth; i++) {
                tocString.append("\t");
            }
            tocString.append(tocReference.getTitle());
            toc.add(tocString.toString());
            getTOCTitles(tocReference.getChildren(), depth + 1, toc);
        }
        return toc;
    }

    private String getChapterTitleFromToc(int chapter, Book mBook) {
        // Is there no easier way to connect a TOCReference
        // to an absolute spine index?
        String title = "";

        List<Resource> targetResources = mBook.getTableOfContents().getAllUniqueResources();
        Resource targetResource = targetResources.get(chapter);

        ArrayList<TOCReference> references = (ArrayList<TOCReference>) mBook.getTableOfContents().getTocReferences();

        for (TOCReference ref : references) {
            if (ref.getResource().equals(targetResource)) {
                return ref.getTitle();
            }
            for (TOCReference childRef : ref.getChildren()) {
                if (childRef.getResource().equals(targetResource)) {
                    return childRef.getTitle();
                }
            }
        }
        return title;
    }


}

File: app/src/main/java/com/example/oceo/speedread/EPubLibUtil.kt
package com.example.oceo.speedread

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import nl.siegmann.epublib.domain.Book
import nl.siegmann.epublib.domain.Resource
import nl.siegmann.epublib.domain.TOCReference
import nl.siegmann.epublib.epub.EpubReader
import nl.siegmann.epublib.service.MediatypeService
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.util.*

class EPubLibUtil {
    private val TAG = "EPubLibUtil"
    private val testBook1Path = "storage/emulated/0/Books/MoonReader/Brandon Sanderson - Oathbringer_ Book Three of the Stormlight Archive-Tor Books (2017).epub"

    private fun getChapterTitleFromToc(chapter: Int, mBook: Book): String {
        // Is there no easier way to connect a TOCReference
        // to an absolute spine index?
        val title = ""
        val targetResources = mBook.tableOfContents.allUniqueResources
        val targetResource = targetResources[chapter]
        val references = mBook.tableOfContents.tocReferences as ArrayList<TOCReference>
        for (ref in references) {
            if (ref.resource == targetResource) {
                return ref.title
            }
            for (childRef in ref.children) {
                if (childRef.resource == targetResource) {
                    return childRef.title
                }
            }
        }
        return title
    }

    companion object {
        private val TAG = "EPubLibUtil"
        var bitmapTypes = arrayOf(MediatypeService.PNG, MediatypeService.GIF, MediatypeService.JPG)

        @JvmStatic
        fun getBook(fName: String?, context: Context): Book? {
            var file: File? = null

            if (!fName!!.contains("asset__")) {
                file = File(fName)
            }

            var book: Book? = null
            try {
                val epubInputStream: InputStream
                if (fName.contains("asset__")) {
                    Log.d(TAG, "making in stream from asset")
                    val fileDescriptor = context.assets.openFd(fName.replace("asset__", ""))
                    epubInputStream = fileDescriptor.createInputStream()
                } else {
                    epubInputStream = FileInputStream(file.toString())
                }

                if (fName.contains(".epub")) {
                    book = EpubReader().readEpub(epubInputStream)
                } else {
                    Log.d("GetBook", "Not an Epub");
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
            return book
        }

        @JvmStatic
        fun getBitmapFromResources(resourcez: List<Resource?>?, imgHref: String, book: Book): Bitmap {
            val resources = book.resources.getResourcesByMediaTypes(bitmapTypes)
            var data = "holder".toByteArray()
            for (ii in resources.indices) {
                val z = resources[ii].href
                // it isnt clear which one will be larger
                // oathbringer chpt 4 is an example of the first condition
                // hyperian prologue is an example of seocnd
                if (imgHref.contains(z) || z.contains(imgHref)) {
//                Log.i("livi", z);
                    try {
                        data = resources[ii].data
                    } catch (e: IOException) {
                        // TODO Auto-generated catch block
                        e.printStackTrace()
                    }
                    break
                }
            }
            //
            return BitmapFactory.decodeByteArray(data, 0, data.size)
        }

        fun getContents(book: Book) {
            val contents = book.contents
        }

        fun exploreResources(book: Book) {
            //TODO
//        Log.d("explorerResources", "start");
            val resource = book.resources
            //        Log.d("explorerResources", "done");
        }

        @JvmStatic
        fun exploreTOC(book: Book): List<TOCReference> {
//        Log.d("explorerTOC", "start");
            val toc = book.tableOfContents
            //        List<Resource> uniqueRes = toc.getAllUniqueResources();
            //        Log.d("explorerTOC", "done");
            return toc.tocReferences
        }

        fun exploreGuide(book: Book) {
//        Log.d("exploreGuide", "start");
            val guide = book.guide
            //        Log.d("exploreGuide", "done");
        }

        fun exploreSpine(book: Book) {
//        Log.d("exploreSpine", "start");
            val spine = book.spine
            //        Log.d("exploreSpine", "done");
        }

        fun unZipEPUB() {
            //TODO
//        FileUtils.unzipFile(mFilePath, Constant.PATH_EPUB + "/" + mFileName);
        }

        @JvmStatic
        fun mapSpineToTOC(spineID: String, tocResList: ArrayList<String>): Int {
            var i = 0
            while (i < tocResList.size) {
                if (spineID == tocResList[i]) {
                    return i
                }
                i++
            }
            return -1
        }

        @JvmStatic
        fun mapTOCToSpine(book: Book, tocID: String): Int {
            val spineRefs = book.spine.spineReferences
            var i = 0
            while (i < spineRefs.size) {
                val spineID = spineRefs[i].resource.id
                if (spineID == tocID) {
                    return i
                }
                i++
            }
            return -1
        }

        @JvmStatic
        fun getTOCResourceIds(tocReferences: List<TOCReference>?, depth: Int, toc: ArrayList<String>): ArrayList<String> {
            if (tocReferences == null) {
                return toc
            }
            for (tocReference in tocReferences) {
                val tocString = StringBuilder()
                for (i in 0 until depth) {
                    tocString.append("\t")
                }
                tocString.append(tocReference.title)
                toc.add(tocReference.resource.id) // resource Id
                getTOCResourceIds(tocReference.children, depth + 1, toc)
            }
            return toc
        }

        @JvmStatic
        fun getTOCTitles(tocReferences: List<TOCReference>?, depth: Int, toc: ArrayList<String>): ArrayList<String> {
            if (tocReferences == null) {
                return toc
            }
            for (tocReference in tocReferences) {
                val tocString = StringBuilder()
                for (i in 0 until depth) {
                    tocString.append("\t")
                }
                tocString.append(tocReference.title)
                toc.add(tocString.toString())
                getTOCTitles(tocReference.children, depth + 1, toc)
            }
            return toc
        }
    }
}